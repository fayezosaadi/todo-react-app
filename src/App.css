.App {
  text-align: center;
  background-color: white;
  margin: 0 20%;
}

.todoListMain .todoListHeader {
  display: flex;
  justify-content: space-between;
  padding: 0 12px;
  background-color: aliceblue;
  height: 68px;
}
.todoListMain .todoListHeader form {
  margin-block-start: 1em;
  margin-block-end: 1em;
  margin-inline-start: 0;
  margin-inline-end: 0;
}
.todoListMain .todoListHeader form label {
  font-size: small;
}

.todoListMain .todoInputForm {
  display: flex;
  justify-content: space-between;
}
.todoListMain .todoInputForm input {
  padding: 10px;
  font-size: 16px;
  border: 2px solid #FFF;
  outline: none;
  width: 100%;
}

.todoListMain .theList {
  padding-inline-start: 0;
  width: 100%;
}
.todoListMain .theList .listItem {
  display: flex;
  justify-content: space-between;
}
.todoListMain .theList form {
  border-bottom: 1px solid darkgrey;
  margin: 10px 25px
}
.todoListMain .theList .listItem button {
  border: none;
  font-weight: bold;
  outline: none;
}
.todoListMain .theList .listItem .listItemText {
  width: 90%;
  display: flex;
}
.todoListMain .theList .listItem .listItemText input {
  margin-right: 13px;
}
