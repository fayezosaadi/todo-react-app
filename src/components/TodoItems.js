import React from 'react';

const TodoItems = ({ items, deleteItem, completeTodo, hide }) => {

  const displayList = (item, i) =>
    <form key={i} style={{ display: hide && item.isCompleted === true ? "none" : "" }}>
      <label style={{ textDecoration: item.isCompleted ? "line-through" : "" }}>
        <div className="listItem">
          <div className="listItemText">
            <input name="isCompleted" type="checkbox" checked={item.isCompleted} onChange={(e) => completeTodo(e, i)} />
            {item.text}
          </div>
          <div>
            <button onClick={(e) => deleteItem(e, item.id)}>x</button>
          </div>
        </div>
        <br />
      </label>
    </form>;

  const listItems = items.map(displayList);

  return <div className="theList">{listItems}</div>
};

export default TodoItems;
