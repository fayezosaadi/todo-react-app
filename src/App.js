import React, { useState, useEffect } from 'react';
import { v4 as uuid } from 'uuid';
import './App.css';
import TodoInput from "./components/TodoInput";
import TodoItems from "./components/TodoItems";
import TodoHeader from "./components/TodoHeader";
const localStorage = require("local-storage");

const currentItemInitialState = { text: '', id: '', isCompleted: false };

const useStateWithLocalStorage = localStorageKey => {
  const [items, setItems] = useState(JSON.parse(localStorage.get(localStorageKey)) || []);

  useEffect(() => { localStorage.set('items', JSON.stringify(items)) }, [items]);

  return [items, setItems];
};

const App = () => {
  const [ items, setItems ] = useStateWithLocalStorage('items');
  const [ hide, setHide ] = useState(false );
  const [ currentItem, setCurrentItem ] = useState(currentItemInitialState);

  const handleInput = e => {
    const { target: { value } } = e;
    setCurrentItem(item => ({ ...item, text: value, id: uuid() }));
};

  const addItem = e => {
    e.preventDefault();

    if (currentItem.text) {
      setItems(itemsList => [...itemsList, currentItem]);
      setCurrentItem(currentItemInitialState);
    }
  };

  const deleteItem = (e, id) => {
    e.preventDefault();

    setItems(items.filter(item => item.id !== id))
  };

  const completeTodo = (e, i) => {
    const { target: { type, checked, value, name } } = e;
    const newTodos = [...items];
    newTodos[i][name] = type === 'checkbox' ? checked : value;
    setItems(newTodos)
  };

  const hideCompleted = (e) => {
    const { target: { type, checked, value } } = e;
    setHide(type === 'checkbox' ? checked : value)
  };

  return (
    <div className="App">
      <div className="todoListMain">
        <TodoHeader hide={hide} hideCompleted={hideCompleted} />
        <TodoInput addItem={addItem} handleInput={handleInput} currentItem={currentItem} />
        <TodoItems items={items} deleteItem={deleteItem} completeTodo={completeTodo} hide={hide} />
      </div>
    </div>
  );
};

export default App;
