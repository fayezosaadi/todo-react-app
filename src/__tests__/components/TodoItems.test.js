import React from "react";
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import TodoItems from "../../components/TodoItems";
import App from "../../App";

const mockItem1 = { text: 'First Task', id: 1, isCompleted: false };
const mockItem2 = { text: 'Second Task', id: 2, isCompleted: false };

describe('<TodoItems />',  () => {

  it('should render', () => {
    render(<TodoItems items={[mockItem1]}/>)
  });

  it('should render and match the snapshot', () => {
    const component = renderer.create(<TodoItems items={[mockItem1]}/>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render all items', () => {
    const { getByText } = render(<TodoItems items={[mockItem1, mockItem2]} />);
    const text1 = getByText('First Task');
    const text2 = getByText('Second Task');
    expect(text1).toBeInTheDocument();
    expect(text2).toBeInTheDocument();
  });

  it('should hide item when isCompleted and hide are true', () => {
    render(<TodoItems hide={true} items={[{...mockItem2, isCompleted: true }]}/>);
    expect(screen.queryByText('Second Task')).toBeNull()
  });

  it('should delete an item', () => {
    const { getByText } = render(<App itemsInitialState={[ mockItem1 ]}/>);
    expect(screen.queryByText('First Task')).toBeTruthy();
    fireEvent.click(getByText(/x/i));
    expect(screen.queryByText('First Task')).toBeNull()
  });

});
