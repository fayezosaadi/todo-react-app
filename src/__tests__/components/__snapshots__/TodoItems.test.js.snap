// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItems /> should render and match the snapshot 1`] = `
<div
  className="theList"
>
  <form>
    <label
      style={
        Object {
          "textDecoration": "",
        }
      }
    >
      <div
        className="listItem"
      >
        <div
          className="listItemText"
        >
          <input
            checked={false}
            name="isCompleted"
            onChange={[Function]}
            type="checkbox"
          />
          First Task
        </div>
        <div>
          <button
            onClick={[Function]}
          >
            x
          </button>
        </div>
      </div>
      <br />
    </label>
  </form>
</div>
`;
